### DISCUSSION TITLE
Java agent is not working in Docker

### DISCUSSION THREAD

**Discussion by haider665** [apache/skywalking#12274] [Q&A] [answered]
I have a very simple spring boot web application. I have added apache skywalking java agent jar which is working fine. But it's not working in docker.

I tried to follow [Java Agent Doc](https://skywalking.apache.org/docs/skywalking-java/next/en/setup/service-agent/java-agent/containerization/#docker) 

Here is my Dockerfile

`
FROM apache/skywalking-java-agent:9.1.0-java17
FROM openjdk:17-jdk
COPY build/libs/*SNAPSHOT.jar app.jar
ENTRYPOINT ["java", "-Dskywalking.agent.service_name=product",  "-Dserver.port=8081", "-jar", "app.jar"]
`

I have checked, there is no log releted to java agent

**Comment by wu-sheng** [comment]
https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/containerization/#docker

**Reply by wu-sheng** [reply]
Or you can do anything you like once you added agent folder and `-javaagent to the JVM.

Docker has nothing specific for Java agent.

**Reply by haider665** [reply]
according to this documentation, it is mentioned that " **you don’t need to care about the Java options to enable SkyWalking agent, it should be adopted automatically"**. Can you please explain this statement?

If I have to add agent folder and pass j-javaagent argument, then '**FROM apache/skywalking-java-agent:8.5.0-jdk8**' what is the usage of this image?

**Reply by wu-sheng** [reply]
This matters, https://github.com/apache/skywalking-java/blob/f736b371d3c92793a25ffa15f6c719c5fe624b82/Dockerfile#L31

This is what standard JVM image provided. You could read the docker file by yourself.

**Reply by haider665** [reply]
I have added the agent folder in my project. Here is my Dockerfile -

`FROM openjdk:17-oracle`
`EXPOSE 8082`
`COPY skywalking-agent /skywalking-agent`
`ENTRYPOINT ["java", "-Dskywalking.agent.service_name=blog", "-Dserver.port=8082", "-javaagent:/skywalking-agent/skywalking-agent.jar", "-jar", "app.jar"]`

Here is my docker compose file -

version: '3.8'
services:
  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2
    container_name: elasticsearch
    ports:
      - "9200:9200"
    healthcheck:
      test: [ "CMD-SHELL", "curl --silent --fail localhost:9200/_cluster/health || exit 1" ]
      interval: 30s
      timeout: 10s
      retries: 3
      start_period: 10s
    environment:
      - discovery.type=single-node
      - bootstrap.memory_lock=true
      - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
    ulimits:
      memlock:
        soft: -1
        hard: -1

  oap:
    image: apache/skywalking-oap-server:10.0.0
    container_name: oap
    depends_on:
      elasticsearch:
        condition: service_healthy
    links:
      - elasticsearch
    ports:
      - "11800:11800"
      - "12800:12800"
    healthcheck:
      test: ["CMD", "bash", "-c", "cat < /dev/null > /dev/tcp/127.0.0.1/11800"]
      interval: 30s
      timeout: 10s
      retries: 3
      start_period: 10s
    environment:
      SW_STORAGE: elasticsearch
      SW_STORAGE_ES_CLUSTER_NODES: elasticsearch:9200
      SW_HEALTH_CHECKER: default
      SW_TELEMETRY: prometheus
      JAVA_OPTS: "-Xms2048m -Xmx2048m"

  ui:
    image: apache/skywalking-ui:10.0.0
    container_name: ui
    depends_on:
      oap:
        condition: service_healthy
    links:
      - oap
    ports:
      - "8080:8080"
    environment:
      SW_OAP_ADDRESS: http://oap:12800
      SW_ZIPKIN_ADDRESS: http://oap:9412

  blog:
    image: haider665/blog:v3
    container_name: blog
    depends_on:
      oap:
        condition: service_healthy
    links:
      - oap
    ports:
      - "8082:8082"
    environment:
      SW_AGENT_COLLECTOR_BACKEND_SERVICES: http://oap:11800
      SW_AGENT_NAME: blog


Unfortunately I am getting this error "**ERROR 2024-06-04 09:21:11.478 SkywalkingAgent-12-GRPCChannelManager-0 GRPCChannelManager : Create channel to  fail. 
java.lang.NullPointerException: Cannot invoke "org.apache.skywalking.apm.agent.core.remote.GRPCChannel.isConnected(boolean)" because "this.managedChannel" is null**"

**Reply by haider665** [reply]
@wu-sheng any idea about this error? I am using 9.2.0 version of java agent

**Reply by wu-sheng** [reply]
I have no idea. Didn't face this before.

**Reply by Superskyyy** [reply, marked as answer]
> @wu-sheng any idea about this error? I am using 9.2.0 version of java agent

Remove the http:// before your oap:port in docker compose `blog` service and try again

**Reply by haider665** [reply]
@Superskyyy it worked, thanks a lot. But why it didn't worked before?

**Reply by kezhenxu94** [reply]
> @Superskyyy it worked, thanks a lot. But why it didn't worked before?

Because you added http protocol prefix to a port that is not http protocol.
